### PR TITLE
chore: lifecycle metadata

### PR DESCRIPTION
**Pull request by jrasm91**
Attempt to track endpoint and dto metadata - when it was added and deprecated

Usage:

```ts
// add a new endpoint
@EndpointLifecycle({ addedAt: 'v1.100.0' })

// deprecate an endpoint
@EndpointLifecycle({ addedAt: '1.100.0', deprecatedAt: '1.103.0' }

// add a new property
@PropertyLifecycle({ addedAt: 'v1.100.0' })

// deprecate a property
@PropertyLifecycle({ addedAt: '1.100.0', deprecatedAt: '1.103.0' }
```

open-api-spec.json
```json
"x-immich-lifecycle": {
  "addedAt": "1.100.0"
  "deprecatedAt": "1.103.0"
}
```

server npm command for a report
```
npm run lifecycle

Changes in NEXT_RELEASE
---------------------
- Deprecated GET /user/info/{id} (getUserById)
- Deprecated PartnerResponseDto.quotaSizeInBytes
- Deprecated UserResponseDto.quotaSizeInBytes

Changes in v1.102.0
---------------------
- Deprecated AddUsersDto.sharedUserIds
- Deprecated AlbumResponseDto.sharedUsers

Changes in 1.100.0
---------------------
- Deprecated GET /assets (searchAssets)

Changes in v1.100.0
---------------------
- Added      GET /user/info/{id} (getUserById)
- Deprecated MemoryLaneResponseDto.title
- Deprecated MetadataSearchDto.resizePath
- Deprecated MetadataSearchDto.webpPath

Changes in v1.90.0
---------------------
- Deprecated GET /search (search)

Changes in v1.80.0
---------------------
- Added      GET /search (search)
```

TODO:
- [ ] Update pump-version.sh to replace NEXT_RELEASE and regenerate the sdks (future PR)

### CODE REVIEW

**Comment by jrasm91** [comment]
Not sure how all the automation will work, but there are several endpoints that need to be deprecated. I think we should maybe just merge this as is so we at least have some information to use moving forward. We can always figure out and add automation in the future.

File: server/src/utils/lifecycle.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+    if (propertySchema.description?.startsWith('This property was added in')) {
+      const addedAt = propertySchema.description.replace('This property was added in', '').trim();
+      trackVersion(addedAt).newProperties.push({ schema: schemaName, property: propertyName });
+    }
+  }
+}
+
+const sortedVersions = Object.keys(metadata).sort((a, b) => {
+  if (a === 'NEXT_RELEASE') {
```
Comment (mertalev) [inline server/src/utils/lifecycle.ts:63]: `NEXT_RELEASE` could be a constant.

File: server/src/utils/lifecycle.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+  if (b === 'NEXT_RELEASE') {
+    return 1;
+  }
+
+  const versionA = Version.fromString(a);
+  const versionB = Version.fromString(b);
+  return versionB.compareTo(versionA);
+});
+
+if (sortedVersions.length > 0) {
```
Comment (mertalev) [inline server/src/utils/lifecycle.ts:76]: This if condition is redundant.

File: server/src/utils/lifecycle.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    if (addedAt) {
+      trackVersion(addedAt).newEndpoints.push({ url, method, endpoint });
+    }
+  }
+}
+
+for (const [schemaName, schema] of Object.entries(spec.components?.schemas || {})) {
+  for (const [propertyName, property] of Object.entries((schema as SchemaObject).properties || {})) {
+    const propertySchema = property as SchemaObject;
+    if (propertySchema.description?.startsWith('This property was deprecated in')) {
```
Comment (mertalev) [inline server/src/utils/lifecycle.ts:50]: `This property was deprecated in` and `This property was added in` can also be constants and used in the decorator.

**Review by mertalev — APPROVED** [review]

File: server/src/decorators.ts [hunk 4]
```diff
@@ -128,3 +129,32 @@ export const GenerateSql = (...options: GenerateSqlQueries[]) => SetMetadata(GEN
...
+    );
+  }
+
+  return applyDecorators(...decorators);
+};
+
+export const PropertyLifecycle = ({ addedAt, deprecatedAt }: LifecycleMetadata) => {
+  const decorators: PropertyDecorator[] = [];
+  decorators.push(ApiProperty({ description: `This property was added in ${addedAt}` }));
+  ApiProperty({ deprecated: true, description: `This property was deprecated in ${deprecatedAt}` });
```
Comment (danieldietzler) [inline server/src/decorators.ts:154]: This line was a bad copy/merge right?

**Review by danieldietzler — APPROVED** [review]
Nice!

**Review by alextran1502 — APPROVED** [review]

**Pull request merged**